components/tidb_query/src/rpn_expr/types/function.rs

```diff
@@ -186,33 +192,37 @@ impl<E: Evaluator> Evaluator for ArgConstructor<E> {
     ) -> Result<VectorValue> {
         match &args[self.arg_index] {
             RpnStackNode::Scalar { value, .. } => {
-                match_template_evaluable! {
-                    TT, match value {
-                        ScalarValue::TT(v) => {
-                            let new_def = Arg {
-                                arg: ScalarArg(v),
-                                rem: def,
-                            };
-                            self.inner.eval(new_def, ctx, output_rows, args, extra)
-                        }
-                    }
+                if let Some(v) = A::try_borrow_scalar_value(value) {
+                    let new_def = Arg {
+                        arg: ScalarArg(v),
+                        rem: def,
+                    };
+                    self.inner.eval(new_def, ctx, output_rows, args, extra)
+                } else {
+                    // TODO: Should print fn name
+                    panic!(
+                        "Cannot apply {:?} at {} argument postion",
+                        value, self.arg_index
+                    );
                 }
             }
             RpnStackNode::Vector { value, .. } => {
                 let logical_rows = value.logical_rows();
-                match_template_evaluable! {
-                    TT, match value.as_ref() {
-                        VectorValue::TT(ref v) => {
-                            let new_def = Arg {
-                                arg: VectorArg {
-                                    physical_col: v,
-                                    logical_rows,
-                                },
-                                rem: def,
-                            };
-                            self.inner.eval(new_def, ctx, output_rows, args, extra)
-                        }
-                    }
+                if let Some(v) = A::try_borrow_vector_value(value.as_ref()) {
+                    let new_def = Arg {
+                        arg: VectorArg {
+                            physical_col: v,
+                            logical_rows,
+                        },
+                        rem: def,
+                    };
+                    self.inner.eval(new_def, ctx, output_rows, args, extra)
+                } else {
+                    // TODO: Should print fn name
+                    panic!(
```

Comment:
Should we panic here @breeswish

Reply:
> Excellent work!
> 
> How about the time of incremental _debug_ compiling tidb_query itself? This is also a very common case.

It's `10s` to `5s` in debug mode with incremental compilation on IDC.

Reply:
It seems to be intended behavior:

https://github.com/tikv/tikv/blob/a55605294bbfeb21ec9d92df89afee01b57690bd/components/tidb_query/src/rpn_expr/types/expr_eval.rs#L202